### PR TITLE
Upgrade to Material UI v4

### LINKED ISSUE #895: [UI] Migrate to material-ui v4

### PR DESCRIPTION
**Pull request by arku**
Closes #895

I followed the steps described in https://material-ui.com/guides/migration-v3. I couldn't get the stuff mentioned under https://material-ui.com/guides/migration-v3/#list though.

@helfi92 How do we go about testing visual regressions (if there are any)?

### CODE REVIEW

File: ui/package.json [hunk 1]
```diff
@@ -75,6 +76,7 @@
     "react-apollo": "^2.5.5",
     "react-copy-to-clipboard": "^5.0.1",
     "react-dom": "^16.8.6",
+    "react-event-listener": "^0.6.6",
```
Comment (arku) [inline ui/package.json]: @helfi92 I had to install this package (dependency of `material-ui-pickers`) as I was running into `Can't resolve 'react-event-listener'` error when running the app. Ideally, we should migrate from `material-ui-pickers` to `@material-ui/pickers` v3.

File: ui/src/views/Documentation/components/Anchor/index.jsx [hunk 2]
```diff
@@ -9,29 +9,33 @@ const useStyles = withStyles(theme => ({
     ...theme.mixins.link,
   },
 }));
+// Ref: https://material-ui.com/guides/composition/#caveat-with-refs
+// eslint-disable-next-line react/display-name
```
Comment (arku) [inline ui/src/views/Documentation/components/Anchor/index.jsx]: To fix the lint rule, I should use a normal `function` as the callback but I am unable to do so as we have another lint rule which requires arrow functions as callbacks and not normal functions.

**Comment by helfi92** [comment]
Wow, this is great. I will review soon. This is going to be so awesome once merged :)

File: ui/package.json [hunk 3]
```diff
@@ -17,9 +17,10 @@
     "create:fragment-matcher": "dotenv node src/scripts/queryServer"
   },
   "dependencies": {
-    "@material-ui/core": "^3.9.3",
+    "@material-ui/core": "^4.2.0",
     "@material-ui/icons": "^3.0.2",
```
Comment (helfi92) [inline ui/package.json]: We could also bump `@material-ui/icons`.

File: ui/package.json [hunk 4]
```diff
@@ -17,9 +17,10 @@
...
     "@material-ui/lab": "3.0.0-alpha.30",
```
Comment (helfi92) [inline ui/package.json]: `@material-ui/lab` could probably also be bumped to the alpha version of v4. We use a couple of components from there. I hope there isn't anything critical to change. The CHANGELOG should be able to say what changed.

File: ui/package.json (hunk 1 above)
Comment (helfi92) [reply ui/package.json]: Agreed we should migrate. I think what's blocking us from migrating to the v3 version of the pickers package is https://github.com/taskcluster/taskcluster/issues/705.

File: ui/src/views/AwsProvisioner/ViewAwsWorkerType/index.jsx [hunk 5]
```diff
@@ -125,7 +125,7 @@ export default class ViewAwsWorkerType extends Component {
     return (
       <Dashboard title={`AWS Provisioner ${workerType}`}>
         <ErrorPanel fixed error={this.state.error || error} />
-        <Tabs fullWidth value={currentTab} onChange={this.handleTabChange}>
+        <Tabs value={currentTab} onChange={this.handleTabChange}>
```
Comment (helfi92) [inline ui/src/views/AwsProvisioner/ViewAwsWorkerType/index.jsx]: What was the reason for removing `fullWidth`?

**Comment by helfi92** [comment]
> @helfi92 How do we go about testing visual regressions (if there are any)?

We will have to do this manually for now.

File: ui/src/views/AwsProvisioner/ViewAwsWorkerType/index.jsx (hunk 5 above)
Comment (arku) [reply ui/src/views/AwsProvisioner/ViewAwsWorkerType/index.jsx]: @helfi92 The deprecated `fullWidth` prop is removed in v4. 

Ref: https://material-ui.com/guides/migration-v3/#tabs

**Comment by arku** [comment]
@helfi92 Thanks for catching the regressions. Let me fix those.

**Comment by helfi92** [comment]
How is it going @arku? Given that this pull-request affects multiples files, the rebase might be painful if we wait too long. There's no rush but just wanted to bring this up so that we can prevent having to do a painful rebase :)

**Comment by arku** [comment]
@helfi92 My bad, forgot this one.

Let me try fixing the regressions.

**Comment by arku** [comment]
@helfi92 I have spent more time on this than I wanted to and it's time to get some help, I guess. Here are my observations regarding the regressions.

### Table row height

Notes from Migration docs:
> [TableRow] Remove the fixed height CSS property. The cell height is computed by the browser using the padding and line-height.

The table rows no longer have a fixed height of 48px. But this is not the only issue. For instance, let's consider the [`HookLastFiredTable`](https://github.com/taskcluster/taskcluster/blob/master/ui/src/components/HookLastFiredTable/index.jsx) component. The [`TableCell`](https://github.com/taskcluster/taskcluster/blob/master/ui/src/components/HookLastFiredTable/index.jsx#L36) component  has a padding of `14px 40px 14px 16px` while the same from v3 has a padding of `4px 56px 4px 24px`. And the [`ListItemText`](https://github.com/taskcluster/taskcluster/blob/master/ui/src/components/HookLastFiredTable/index.jsx#L42) component has a top and bottom margin of `4px` while in v3, it has no margin. I noticed the `div` rendered by the `ListItemText` component has a CSS class of `MuiListItemText-dense` (which should set the margin to 0) but the CSS styles have no reference to it :confused: 

To make the sure the table looks the same as in v3, we have to add styles to the `TableCell` and `ListItemText`. This would mean we have to do the same for all `TableCell`s in all `Table` components and the same for `ListItemText` components. I am not sure if this is the right way to go about it though. What do you think?

### Font size

Notes from Migration docs:

> [Typography] Change the default variant from body2 to body1. A font size of 16px is a better default than 14px. Bootstrap, material.io, and even our documentation use 16px as a default font size. 14px like Ant Design uses is understandable, as Chinese users have a different alphabet. We recommend 12px as the default font size for Japanese.

I have updated the variants for the `Typography` components. I noticed a style of `font-size: 0.875rem;` added to `body` but I don't understand why it is injected. We should figure out why it is injected and remove it maybe?

### Color

A style of `rgba(255, 255, 255, 0.7)` is added to `html` and `body`.

Upon some digging, it looks like the [`textSecondary`](https://github.com/taskcluster/taskcluster/blob/master/ui/src/theme.js#L39) color from our theme is what's used for above style. It should be the `textPrimary` color though.

### Drawer

I noticed a spacing issue in the docs site after the upgrade but I didn't have any luck fixing it :disappointed: 

docs_spacing_issue

Do you mind helping me out with these issues? Once these are addressed, we have to check all the pages manually.

**Comment by helfi92** [comment]
Thank you for the all of the information 🥇 I plan on taking a look as soon as I get the chance.

**Comment by arku** [comment]
@helfi92 Sounds good, thanks!

**Comment by helfi92** [comment]
Continues in https://github.com/taskcluster/taskcluster/pull/1729.

**Pull request closed without merging**